File: CS/ObjectDataSourceInCode/Sample4.cs
using DevExpress.XtraReports.UI;
using System.Collections.Generic;
using System.Linq;

namespace ObjectDataSourceInCode.Sample4
{

    // This sample demonstrates how to bind a report to the data that an object's method exposes. This data is filtered by the method's parameters.
    public class CodeSample {
        public static XtraReport CreateReport() {

            var objectDS = new DevExpress.DataAccess.ObjectBinding.ObjectDataSource() {
                Name = "ObjectDataSource1",
                Parameters = {
                    new DevExpress.DataAccess.ObjectBinding.Parameter(
                        "parameter1",
                        typeof(DevExpress.DataAccess.Expression), 
                        new DevExpress.DataAccess.Expression("?reportParameter1", typeof(int))),
                    
                    new DevExpress.DataAccess.ObjectBinding.Parameter(
                        "parameter2",
                        typeof(int),
                        5)
                },
                DataMember = "GetData",
                DataSource = typeof(DataModel),
            };

            var report = new XtraReport() {
                Bands = {
                    new TopMarginBand() { Name = "TopMargin" },
                    new DetailBand() {
                        Name = "Detail",
                        Controls = {
                            new XRLabel() { 
                                Name = "label1",
                                ExpressionBindings = { new ExpressionBinding("Text", "[Name]") },
                                Size = new System.Drawing.Size(100, 23)
                            },
                        }
                    },
                    new BottomMarginBand() { Name = "BottomMargin" },
                },
                Parameters = {
                    new DevExpress.XtraReports.Parameters.Parameter() {
                        Name = "reportParameter1",
                        Type = typeof(int),
                        Value = 10
                    }
                },
                DataSource = objectDS
            };

            return report;
        }
    }
    public class DataModel {
        public static List<DataItem> GetData(int parameter1, int parameter2) {
            return new List<DataItem>(
                Enumerable.Range(parameter1, parameter2).Select(item => new DataItem() { Name = "Name " + item })
            );
        }
    }
    public class DataItem {
        public string Name { get; set; }
    }
}

File: VB/ObjectDataSourceInCode/Sample4.vb
Imports DevExpress.XtraReports.UI
Imports System.Collections.Generic
Imports System.Linq

Namespace ObjectDataSourceInCode.Sample4

    ' This sample demonstrates how to bind a report to the data that an object's method exposes. This data is filtered by the method's parameters.
    Public Class CodeSample
        Public Shared Function CreateReport() As XtraReport

            Dim objectDS = New DevExpress.DataAccess.ObjectBinding.ObjectDataSource() With { _
                .Name = "ObjectDataSource1", _
                .Parameters = { _
                    New DevExpress.DataAccess.ObjectBinding.Parameter("parameter1", GetType(DevExpress.DataAccess.Expression), New DevExpress.DataAccess.Expression("?reportParameter1", GetType(Integer))), _
                    New DevExpress.DataAccess.ObjectBinding.Parameter("parameter2", GetType(Integer), 5) _
                }, _
                .DataMember = "GetData", _
                .DataSource = GetType(DataModel) _
            }

            Dim report = New XtraReport() With { _
                .Bands = { _
                    New TopMarginBand() With {.Name = "TopMargin"}, _
                    New DetailBand() With { _
                        .Name = "Detail", _
                        .Controls = { _
                            New XRLabel() With { _
                                .Name = "label1", .ExpressionBindings = { New ExpressionBinding("Text", "[Name]") }, _
                                .Size = New System.Drawing.Size(100, 23) _
                            } _
                        } _
                        _
                    }, _
                    New BottomMarginBand() With {.Name = "BottomMargin"} _
                }, _
                .Parameters = { _
                    New DevExpress.XtraReports.Parameters.Parameter() With { _
                        .Name = "reportParameter1", _
                        .Type = GetType(Integer), _
                        .Value = 10 _
                    } _
                }, _
                _
                .DataSource = objectDS _
            }

            Return report
        End Function
    End Class
    Public Class DataModel
        Public Shared Function GetData(ByVal parameter1 As Integer, ByVal parameter2 As Integer) As List(Of DataItem)
            Return New List(Of DataItem)(Enumerable.Range(parameter1, parameter2).Select(Function(item) New DataItem() With {.Name = "Name " & item}))
        End Function
    End Class
    Public Class DataItem
        Public Property Name() As String
    End Class
End Namespace
